### PR TITLE
Add promise support to all public functions

### PR DESCRIPTION
**Pull request by dmihal**
Discussed in #252

### CODE REVIEW

**Comment by holgerd77** [comment]
Whew. What a PR! 👍 🍰 😄

**Comment by holgerd77** [comment]
@dmihal The Circle tests can be ignored for now, they are not working atm (at least they have to be re-triggered to run through), passing travis is essential.

**Comment by dmihal** [comment]
@holgerd77 It looks bigger than it is, since it re-indents most of the public API functions.

[Here's the diff with whitespace ignored](https://github.com/ethereumjs/ethereumjs-vm/pull/253/files?w=1)

**Comment by holgerd77** [comment]
@dmihal Ah, didn’t realize this on first look. 😄 

@yann300 Does this PR create any potential problems for remix?

**Comment by holgerd77** [comment]
@dmihal Ah, tests passing, cool!

File: lib/index.js [hunk 1]
```diff
@@ -84,9 +85,17 @@ VM.prototype.copy = function () {
  * Loads precompiled contracts into the state
  */
 VM.prototype.loadCompiled = function (address, src, cb) {
-  this.trie.db.put(address, src, cb)
+  var self = this
+  return new Promise(function (resolve, reject) {
+    cb = cb || function (err, result) { err ? reject(err) : resolve(result) }
```
Comment (axic) [inline lib/index.js]: I don't think it is a good idea to return a promise which will never resolve - that is the case when a callback is passed (e.g. the old-world compatibility mode).

Wouldn't it make sense keeping the old interface unchanged and adding a promise API in front (or the other way around).

Also I'd be supportive internally moving to promises, which is a lot of work, and that would mean that the call back interface would need become the wrapper API.

Comment (holgerd77) [reply lib/index.js]: Can't we just put the switch in an if statement to have a more clean implementation with not returning an unresolved promise and at the same time keep the original idea and not have an additional/second API?

```javascript
VM.prototype._populateCache = function (addresses, cb) {
  self.stateManager.warmCache(addresses, cb)
}

VM.prototype.populateCache = function (addresses, cb) {
  if (!cb) {
    var self = this
    return new Promise(function (resolve, reject) {
      cb = function (err, result) { err ? reject(err) : resolve(result) }
      self._populateCache(addresses, cb)
    })
  } else {
    self._populateCache(addresses, cb)
  }
}
```

**Comment by dmihal** [comment]
Going to close this, #273 is a much cleaner change

**Pull request closed without merging**